Development/nmos/control_protocol_utils.cpp

```diff
@@ -0,0 +1,648 @@
+        {
+            auto& members = nmos::fields::nc::members(resource.data);
+
+            auto members_found = find_members_by_matching_class_id(members);
+            for (const auto& member : members_found)
+            {
+                web::json::push_back(descriptors, member);
+            }
+
+            if (recurse)
+            {
+                // do class_id match on all NcBlock(s)
+                for (const auto& member : members)
+                {
+                    if (is_nc_block(nmos::details::parse_nc_class_id(nmos::fields::nc::class_id(member))))
+                    {
+                        // get resource based on the oid
+                        const auto& oid = nmos::fields::nc::oid(member);
+                        const auto& found = find_resource(resources, utility::s2us(std::to_string(oid)));
+                        if (resources.end() != found)
+                        {
+                            find_members_by_class_id(resources, *found, class_id_, include_derived, recurse, descriptors);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // push a control protocol resource into other control protocol NcBlock resource
+    void push_back(control_protocol_resource& nc_block_resource, const control_protocol_resource& resource)
+    {
+        using web::json::value;
+
+        auto& parent = nc_block_resource.data;
+        const auto& child = resource.data;
+
+        if (!is_nc_block(details::parse_nc_class_id(nmos::fields::nc::class_id(parent)))) throw std::logic_error("non-NcBlock cannot be nested");
+
+        web::json::push_back(parent[nmos::fields::nc::members],
+            details::make_nc_block_member_descriptor(nmos::fields::description(child), nmos::fields::nc::role(child), nmos::fields::nc::oid(child), nmos::fields::nc::constant_oid(child), details::parse_nc_class_id(nmos::fields::nc::class_id(child)), nmos::fields::nc::user_label(child), nmos::fields::nc::oid(parent)));
+
+        nc_block_resource.resources.push_back(resource);
+    }
+
+    // modify a control protocol resource, and insert notification event to all subscriptions
+    bool modify_control_protocol_resource(resources& resources, const id& id, std::function<void(resource&)> modifier, const web::json::value& notification_event)
+    {
+        auto found = resources.find(id);
+        if (resources.end() == found || !found->has_data()) return false;
+
+        auto pre = found->data;
+
+        // "If an exception is thrown by some user-provided operation, then the element pointed to by position is erased."
+        // This seems too surprising, despite the fact that it means that a modification may have been partially completed,
+        // so capture and rethrow.
+        // See https://www.boost.org/doc/libs/1_68_0/libs/multi_index/doc/reference/ord_indices.html#modify
+        std::exception_ptr modifier_exception;
+
```

Comment:
Should there also be a write lock?

Reply:
The write lock is already in the outer function to protect it.

Will add a note in the function to remind the user to set the write lock in the outer function.